src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,889 @@
+    _instance->_sampler->on_javathread_create(thread);
+  }
+}
+
+void JfrCPUTimeThreadSampling::on_javathread_terminate(JavaThread *thread) {
+  if (_instance != nullptr && _instance->_sampler != nullptr) {
+    _instance->_sampler->on_javathread_terminate(thread);
+  }
+}
+
+void handle_timer_signal(int signo, siginfo_t* info, void* context) {
+  assert(_instance != nullptr, "invariant");
+  _instance->handle_timer_signal(context);
+}
+
+
+void JfrCPUTimeThreadSampling::handle_timer_signal(void* context) {
+  assert(_sampler != nullptr, "invariant");
+  if (Atomic::load(&_sampler->_stop_signals)) {
+    return;
+  }
+  Atomic::inc(&_sampler->_active_signal_handlers);
+  _sampler->handle_timer_signal(context);
+  Atomic::dec(&_sampler->_active_signal_handlers);
+}
+
+void JfrCPUTimeThreadSampler::handle_timer_signal(void* context) {
+  JavaThread* jt = get_java_thread_if_valid();
+  if (jt == nullptr) {
+    return;
+  }
+  JfrCPUTimeTrace* trace = this->_queues.fresh().dequeue();
+  if (trace != nullptr) {
+    trace->record_trace(jt, context, get_sampling_period());
+    this->_queues.filled().enqueue(trace);
+  } else {
+    Atomic::inc(&_ignore_because_queue_full);
+    Atomic::inc(&_ignore_because_queue_full_sum);
+  }
+}
+
+const int SIG = SIGPROF;
+
+void JfrCPUTimeThreadSampler::set_timer_time(timer_t timerid, int64_t period_nanos) {
+  struct itimerspec its;
+  if (period_nanos == 0) {
+    its.it_interval.tv_sec = 0;
+    its.it_interval.tv_nsec = 0;
+  } else {
+    its.it_interval.tv_sec = period_nanos / NANOSECS_PER_SEC;
+    its.it_interval.tv_nsec = period_nanos % NANOSECS_PER_SEC;
+  }
+  its.it_value = its.it_interval;
+  if (timer_settime(timerid, 0, &its, NULL) == -1) {
+    warning("Failed to set timer for thread sampling: %s", os::strerror(os::get_last_error()));
+  }
+}
+
+bool JfrCPUTimeThreadSampler::create_timer_for_thread(JavaThread* thread, timer_t& timerid) {
+  if (thread->osthread() == nullptr || thread->osthread()->thread_id() == 0){
+    return false;
+  }
+  timer_t t;
+  OSThread::thread_id_t tid = thread->osthread()->thread_id();
+  struct sigevent sev;
+  sev.sigev_notify = SIGEV_THREAD_ID;
+  sev.sigev_signo = SIG;
+  sev.sigev_value.sival_ptr = &t;
+  ((int*)&sev.sigev_notify)[1] = tid;
+  clockid_t clock;
+  int err = pthread_getcpuclockid(thread->osthread()->pthread_id(), &clock);
+  if (err != 0) {
+    log_error(jfr)("Failed to get clock for thread sampling: %s", os::strerror(err));
+    return false;
+  }
+  if (timer_create(clock, &sev, &t) < 0) {
+    return false;
+  }
+  int64_t period = get_sampling_period();
+  if (period != 0) {
+    set_timer_time(t, period);
+  }
+  timerid = t;
+  return true;
+}
+
+void JfrCPUTimeThreadSampler::init_timers() {
+  // install sig handler for sig
+  PosixSignals::install_generic_signal_handler(SIG, (void*)::handle_timer_signal);
+
+  // create timers for all existing threads
+  MutexLocker tlock(Threads_lock);
```

Comment:
When exactly is this called? The direct use of the Threads_lock raises concerns. Should this not be done at a global safepoint?

Reply:
This is called transitively from the Java code whenever the event is first enabled. So this should lead to a safepoint?

Reply:
AFAIK, we are removing the timers once all recordings with the cpu profiler event are closed. And would re-add them when another recording is started later, right?
In that case, if we force a global safepoint in order to init/remove timers we could get an unexpectly high amount of safepoints, dependent on the external factors starting/stopping/configuring JFR recordings.

@dholmes-ora What exactly is the concern here? I just want to understand if the concern really warrants increasing the number of global safepoints when using JFR with this sampler.

Reply:
@jbachorik if you take out the `Threads_lock` you will delay real safepoints. Direct use of the Threads_lock is not desirable as it can lead to unexpected deadlocks and interferes with safepoints.

If all you want is to create timers for all existing threads then use of a ThreadsListIterator should suffice as a snapshot.

Reply:
Of course you are racing with the execution of those other threads either way. In contrast if you do this at a safepoint you know the exact state of each thread and will know, for example, that the timer was created before the thread could terminate.

Reply:
How would I implement this?

Reply:
Whether the racing is an issue only you can answer. My concern is that using the `Threads_lock` is bad and unnecessary - if you want a stable set of threads then you just need a `ThreadsListIterator`.

Otherwise you'd implement with a safepoint VM operation that you submit to the VMThread.

Reply:
I just looked into it. We only call the methods that take the locks when
- the sampler is first enrolled
- the timer intervals are updated
- the timer is stopped

This shouldn't happen that often in practice, so I think that it is good to have the additional guarantees offered by safepoints.

Reply:
So maybe we do timer init and removal at global safepoints?